### ISSUE TITLE
[IDEA] <$set> widgets can be accidentally picked up by \import. Let's stop that.

### ISSUE THREAD

**Issue by flibbles** [Jermolene/TiddlyWiki5#7997]
Set widgets can get caught by \import. For instance:
```
\procedure myGlobalVariable() ...

<$set name=localVariable value=...>

...
```

`localVariable` will get roped into `$:/tags/Global` as well, which is bizarre because the design of <$set>, as an enclosed widget, implies that its defined variable should only exist between its opening and closing tag.

@Jermolene, you may remember that I actually pointed this issue out a long time ago, but it was decided to keep it in because it might be a useful feature. But I'm bringing it up again, for the following reasons.

1. I don't think anyone knows this exists. Not even power-users. I only do because I revamped the import widget several years back, and I had to make special handling to deal with this case which made import a lot harder to understand. If we remove this, `<$import>` can be refactored to be simpler.
2. The `<$parameters>` widget gives pragma definitions a lot more power. There is still a narrow scope of things you can do with `<$set>` that you can't do with pragma, but that gulf has narrowed considerably.
3. This is **super brittle**. Take this case:
    ```
    \procedure global() ...
    <$set name=otherGlobal value=...>
    <$set name=thirdGlobal value=...>
    ...
    </$set><$/set>
    ```
    This does not work. You need to have two newlines after *each* $set or they won't import. That's because they're getting treated as inline widgets, and `<p>` tags are being introduced before them that curtails importing.

    How about this?
    ```
    \procedure global()
    <$set name=otherGlobal value=... />

    <$set name=thirdGlobal value=.../>

    ```
    Think you can be cute and use self-closing `<$set>` because that makes them more like the pragma? Get out of here. They need to be nested within each other or they don't work.

    How about this?
    ```
    \procedure global()
    <$let
      otherGlobal=...
      thirdGlobal=...>

    ...
    </$let>
    ```
    Think you can use any of the shorter form definers just because that's logical? We don't do that here. `<$set>` and the long-forgotten `<$setvariable>` only.

    The only way these work is if you have nested `<$set>` widgets, each two lines apart from each other or any body content, with no other content except what is inside the most nested element. In other words, you have to know exactly what you're doing to make these work, and I think if people are using them, they're using them by accident.
4. I'm bringing this back up because it came up when diagnosing a Relink bug, and technically if we remove this, I don't have to fix the bug. But I _can_. This isn't really my main reason.

So all that combined, I thought it was worth bringing up this hidden "feature" and inquire whether it should still be kept. I think it should go. I think pragma should be the only way to define globally scoped variables because widgets imply they're self-contained. Also, this isn't documented anywhere.

**Comment by saqimtiaz** [comment]
> 1. I don't think anyone knows this exists. Not even power-users. I only do because I revamped the import widget several years back, and I had to make special handling to deal with this case which made import a lot harder to understand. If we remove this, `<$import>` can be refactored to be simpler.

Irrespective of the merits of the remainder of your argument, this part is not true. I use this quite often and in fact rely on it for some of my wikis, and those wikis have other active users. It has also been discussed as a technique in the community form in the recent past. So if we were to change this, it would most definitely be a backwards compatibility break, and that needs to be acknowledged.

**Comment by flibbles** [comment]
> I use this quite often and in fact rely on it for some of my wikis, and those wikis have other active users.

Okay. Then I guess we'll have to keep it, but how did you even know this feature even existed?

**Comment by saqimtiaz** [comment]
I don't honestly recall how I learnt about it as I have been using it for quite some time. It was the only feasible mechanism at the time for assigning a global variable that needed to be computed from a filter, though the function pragma might offer other possibilities now.

**Comment by flibbles** [comment]
I guess you did just say it's been recently discussed. In those discussions, did anyone bring up how weird and unintuitive this feature is? I mean, it seems obvious to me from the code that it exists as an unintended consequence.

**Comment by saqimtiaz** [comment]
@flibbles the idiosyncrasies were indeed discussed including some brief discussion of the need for better documentation. To be clear, I don't have any personal objection to upgrading my wikis with a workaround if one exists now with `\function` , however it goes to show that it is a backwards compatibility break and others may be impacted as well.

**Comment by flibbles** [comment]
That's true about backward-compatibility. It has to stay if people are *actually discussing this like it's a solution and not black sorcery.* I guess I'll close this and get to work fixing Relink.
